src/nvim/autocmd.c

```diff
@@ -1112,46 +1103,27 @@ int autocmd_register(int64_t id, event_T event, const char *pat, int patlen, int
   return OK;
 }
 
-size_t aucmd_pattern_length(const char *pat)
-  FUNC_ATTR_PURE
+size_t aucmd_span_pattern(const char *pat, const char **start)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
 {
-  if (*pat == NUL) {
-    return 0;
+  // Skip leading commas.
+  while (*pat == ',') {
+    pat++;
   }
 
-  const char *endpat;
-
-  for (; *pat; pat = endpat + 1) {
-    // Find end of the pattern.
-    // Watch out for a comma in braces, like "*.\{obj,o\}".
-    endpat = pat;
-    // ignore single comma
-    if (*endpat == ',') {
-      continue;
+  // Find end of the pattern.
+  // Watch out for a comma in braces, like "*.\{obj,o\}".
+  const char *p = pat;
+  for (int brace_level = 0; *p && (*p != ',' || brace_level || (p > pat && p[-1] == '\\')); p++) {
```

Comment:
I'd prefer initializing `brace_level` before the `for`. It just feels a bit strange to initialize it in the `for` init statement when `p` is initialized before the `for`.
```suggestion
  const char *p = pat;
  int brace_level = 0;
  for (; *p && (*p != ',' || brace_level || (p > pat && p[-1] == '\\')); p++) {
```